spack external find: allow to search by tags

alalazo opened the pull request:
This commit adds an option to the:
```
spack external find
```
command that allows it to search by tags. In this way group of executables with common purposes can be grouped under a single name and a simple command can be used to detect all of them. As an example introduce the "build-tools" tag to search for common development tools on a system.

Example of use:
```console
% spack external find -t build-tools
==> The following specs have been detected on this system and added to /Users/culpo/.spack/packages.yaml
autoconf@2.69  automake@1.16.2  cmake@3.19.1  flex@2.5.35  gmake@3.81  libtool@2.4.6  m4@1.4.6  pkgconf@1.7.3
```

alalazo committed:
spack external find: allow to search by tags

This commit adds an option to the `external find`
command that allows it to search by tags. In this
way group of executables with common purposes can
be grouped under a common name and a simple command
can be used to detect any of them.

As an example introduce the 'dev-tools' tag to
search for common development tools on a system

alalazo commented:
This is just an example, but in the future we may add tags for other purposes for instance:
```console
spack external find -t gpu
```
to detect all the software provided by vendors related to GPUs or anything else that may seem useful.

adamjstewart commented on lib/spack/spack/test/cmd/external.py:
```diff
@@ -242,3 +242,15 @@ def test_new_entries_are_reported_correctly(
+
+
+def test_use_tags_for_detection(mock_executable, mutable_config, monkeypatch):
+    # Prepare an environment to detect a fake cmake
```
You should also add a fake non-dev-tools package to the environment and make sure it does not get detected

alalazo replied:
Done

becker33 commented on lib/spack/spack/cmd/external.py:
```diff
@@ -148,10 +152,21 @@ def _spec_is_valid(spec):
+    packages_to_check = []
+
+    # If tags have been passed, add the corresponding packages
+    if args.tags:
+        for pkg in spack.repo.path.packages_with_tags(*args.tags):
+            packages_to_check.append(spack.repo.get(pkg))
```
I'm not sure this isn't the behavior we want.

What should `spack external find -t dev-tools autoconf openmpi` do? Should it

1. find `autoconf`, because that is the listed package that satisfies the tag
2. find `openmpi` + all packages tagged as dev-tools.

I would lean towards the former.

alalazo replied:
I was leaning towards the latter obviously, but have no strong preference. If tags are considered filters, like you propose, the most common use case would still remain:
```console
$ spack external find -t build-tools 
```
I'll do the change and ping when done

becker33 commented on lib/spack/spack/cmd/external.py:
```diff
@@ -42,6 +42,10 @@ def setup_parser(subparser):
         default=spack.config.default_modify_scope('packages'),
         help="configuration scope to modify")
+    find_parser.add_argument(
+        '-t', '--tags', action='append',
```
This argument should be `--tag` if it only accepts a single tag, `--tags` if it accepts a comma-delimited list

becker33 reviewed (changes requested):
I think the name of the tag you added would be more accurate as `build-tools` than `dev-tools`. I think things like debuggers would go in `dev-tools`, but would not belong in this category.

alalazo committed:
Address code review

Use tags to filter the list of packages rather than
to add packages.

Use --tag (singular) instead of --tags everywhere

Rename dev-tools into build-tools

alalazo committed:
Add a unit test to check filtering with packages

alalazo commented:
@becker33 @adamjstewart This is now 100% covered. I think the failure on codecov tests is just spurious. Let me know if any other change is needed.
